File: LeetCode/53.maximum-subarray.cpp
/*
 * @lc app=leetcode id=53 lang=cpp
 *
 * [53] Maximum Subarray
 */

// @lc code=start

class Solution {
public:
    int maxSubArray(vector<int>& nums) {
         int max_sub = nums[0];
    int currunt_max = 0;


    for (int i = 0; i < nums.size(); i++) { 
        // if curr < 0 return to z;
        currunt_max = max(currunt_max, 0);

        currunt_max = currunt_max + nums[i];

        max_sub = max(max_sub, currunt_max);
    }
    return max_sub;
    }
};
// @lc code=end

File: LeetCode/53.maximum-subarray.py

from typing import List

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        if not nums:
            return -1
        if len(nums) == 1:
            return nums[0]
        curr_max = 0
        final_max = nums[0]
        for num in nums:
            if curr_max < 0:
                curr_max = 0
            curr_max += num
            final_max = max(final_max, num)

        return final_max


s = Solution()
print(s.maxSubArray([-2,-1]))